synapse/rest/client/v2_alpha/shared_rooms.py

```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 Half-Shot
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from synapse.api.errors import Codes, SynapseError
+from synapse.http.servlet import RestServlet
+
+from ._base import client_patterns
+
+logger = logging.getLogger(__name__)
+
+
+class UserSharedRoomsServlet(RestServlet):
+    """
+    GET /uk.half-shot.msc2666/user/shared_rooms/{user_id} HTTP/1.1
+    """
+
+    PATTERNS = client_patterns(
+        "/uk.half-shot.msc2666/user/shared_rooms/(?P<user_id>[^/]*)",
+        releases=(),  # This is an unstable feature
+    )
+
+    def __init__(self, hs):
+        super(UserSharedRoomsServlet, self).__init__()
+        self.auth = hs.get_auth()
+        self.store = hs.get_datastore()
+
+    async def on_GET(self, request, user_id):
+        requester = await self.auth.get_user_by_req(request)
+        if user_id == requester.user.to_string():
+            raise SynapseError(
+                code=400,
+                msg="'user_id' must not be the authenticated user",
+                errcode=Codes.FORBIDDEN,
+            )
+
+        rooms = await self.store.get_rooms_in_common_for_users(
```

Comment:
I'm afraid this method was removed in https://github.com/matrix-org/synapse/pull/8045, as it was unused ever since a [rewrite of the user directory](https://github.com/matrix-org/synapse/pull/4537) occurred to calculate the user directory iteratively instead of doing an expensive calculation per-request.

The [original implementation](https://github.com/matrix-org/synapse/pull/8045/files#diff-a91810f1149e0f4efc0f75a33cdcbefdL678-L701) for this method was rather expensive. From the new user directory code, I see that there are `users_who_share_private_rooms` and `users_in_public_rooms` tables, both of which include remote users. Methods using them can be found here: https://github.com/matrix-org/synapse/blob/f3fe6961b211d898aa347771df598c531fbca90c/synapse/storage/databases/main/user_directory.py#L638-L664

You'll need to add another that gets rooms in common between two users though. Might I also suggest that this method have a cache thrown on it? You can do so by throwing:

```
@cached()
```

on top of the method declaration.

Reply:
I've reimplemented it as best as I can in pure SQL. It looks like it has legs, and a cache.